File: CSES-sheet/Graph/Round-Trip-task-1669/code.cpp
#include <bits/stdc++.h>
using namespace std;

vector<vector<int>> adj;
vector<int> cycle;

bool dfs(int node, int par, vector<bool> &vis, vector<int> &parent) {
    vis[node] = true;
    parent[node] = par;

    for (int child : adj[node]) {

        if (child == par)
            continue;

        if (!vis[child]) {

            if (dfs(child, node, vis, parent))
                return true;

        } else {

            // Back edge found => cycle
            cycle.clear();

            cycle.push_back(child);

            int cur = node;
            while (cur != child) {
                cycle.push_back(cur);
                cur = parent[cur];
            }

            cycle.push_back(child);
            reverse(cycle.begin(), cycle.end());

            return true;
        }
    }

    return false;
}

int main() {
    ios::sync_with_stdio(false);
    cin.tie(nullptr);

    int n, m;
    cin >> n >> m;

    adj.assign(n + 1, vector<int>());

    for (int i = 0; i < m; i++) {
        int u, v;
        cin >> u >> v;
        adj[u].push_back(v);
        adj[v].push_back(u);
    }

    vector<bool> vis(n + 1, false);
    vector<int> parent(n + 1, -1);

    bool found = false;

    for (int i = 1; i <= n; i++) {
        if (!vis[i]) {
            if (dfs(i, -1, vis, parent)) {
                found = true;
                break;
            }
        }
    }

    if (!found) {
        cout << "IMPOSSIBLE\n";
    } else {
        cout << cycle.size() << '\n';
        for (int x : cycle)
            cout << x << " ";
        cout << '\n';
    }

    return 0;
}

File: CSES-sheet/Graph/Round-Trip-task-1669/code.java
//https://cses.fi/problemset/task/1669/
import java.io.*;
import java.util.*;

public class Demo {

    static List<List<Integer>> adj;
    static List<Integer> cycle;

    public static boolean dfs(int node, int par, boolean[] vis, int[] parent) {

        vis[node] = true;
        parent[node] = par;

        for (int child : adj.get(node)) {

            if (child == par)
                continue;

            if (!vis[child]) {

                if (dfs(child, node, vis, parent))
                    return true;

            } else {

                // Back edge found => cycle
                cycle = new ArrayList<>();

                cycle.add(child);

                int cur = node;
                while (cur != child) {
                    cycle.add(cur);
                    cur = parent[cur];
                }

                cycle.add(child);
                Collections.reverse(cycle);

                return true;
            }
        }

        return false;
    }

    public static void solve(Scanner sc) {
        adj = new ArrayList<>();

        int n = sc.nextInt();
        int m = sc.nextInt();

        for (int i = 0; i <= n; i++) {
            adj.add(new ArrayList<>());
        }

        for (int i = 0; i < m; i++) {
            int u = sc.nextInt();
            int v = sc.nextInt();
            adj.get(u).add(v);
            adj.get(v).add(u);
        }

        boolean[] vis = new boolean[n + 1];
        int[] parent = new int[n + 1];
        Arrays.fill(parent, -1);

        cycle = null;

        boolean found = false;

        for (int i = 1; i <= n; i++) {
            if (!vis[i]) {
                if (dfs(i, -1, vis, parent)) {
                    found = true;
                    break;
                }
            }
        }

        if (!found) {
            System.out.println("IMPOSSIBLE");
        } else {
            System.out.println(cycle.size());
            for (int x : cycle)
                System.out.print(x + " ");
        }
    }

    public static void main(String[] args) throws Exception {
        Scanner sc = new Scanner(System.in);
        solve(sc);
        sc.close();
    }
}
